### PR TITLE
Document and comment `Resource` and `ResourceOwner`.

### PR DESCRIPTION
**Pull request by fabiotudone-da**
### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 1]
```diff
@@ -43,12 +67,18 @@ trait Resource[+A] {
...
+    * A nested resource can be flattened.
+    */
   def flatten[B](
       implicit nestedEvidence: A <:< Resource[B],
       executionContext: ExecutionContext,
   ): Resource[B] =
     flatMap(identity[A])
 
+  /**
+    * Just like [[Future]]s, an attempted [[Resource]] computation can chained to a [[Resource]].
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
    * Just like [[Future]]s, an attempted [[Resource]] computation can be transformed.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 2]
```diff
@@ -58,33 +88,40 @@ trait Resource[+A] {
...
   )(implicit executionContext: ExecutionContext): Resource[T] =
     new Resource[T] {
-      private val released: AtomicBoolean = new AtomicBoolean(false)
-      private val releasePromise: Promise[Unit] = Promise()
 
       final lazy val asFuture: Future[T] = future.transformWith {
         case Success(value) => Future.successful(value)
-        case Failure(throwable) => release().flatMap(_ => Future.failed(throwable))
+        case Failure(throwable) =>
+          release().flatMap(_ => Future.failed(throwable)) // Release all if failing
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
          release().flatMap(_ => Future.failed(throwable)) // Release everything on failure
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 3]
```diff
@@ -9,30 +9,54 @@ import scala.collection.generic.CanBuildFrom
...
 trait Resource[+A] {
   self =>
 
+  /**
+    * Every [[Resource]] has an underlying [[Future]] representation.
+    */
   val asFuture: Future[A]
 
+  /**
+    * Every [[Resource]] can be (asynchronously) released.
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
    * Every [[Resource]] can be (asynchronously) released. Releasing a resource will also release
    * all earlier resources constructed via [[flatMap()]] or a `for` comprehension.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 4]
```diff
@@ -58,33 +88,40 @@ trait Resource[+A] {
...
         case Success(value) => Future.successful(value)
-        case Failure(throwable) => release().flatMap(_ => Future.failed(throwable))
+        case Failure(throwable) =>
+          release().flatMap(_ => Future.failed(throwable)) // Release all if failing
       }
 
+      private val released: AtomicBoolean = new AtomicBoolean(false) // Short-circuits to a promise
+      private val releasePromise: Promise[Unit] = Promise() // Will be the release return handle
+
+      // Explicit release must call both main and nested releases
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]: I think you can delete this comment and make it part of the Scaladoc of the trait.

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 5]
```diff
@@ -58,33 +88,40 @@ trait Resource[+A] {
...
       def release(): Future[Unit] =
-        if (released.compareAndSet(false, true))
+        if (released.compareAndSet(false, true)) // Atomically: if not yet released, set released
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
        // If `release` is called twice, we wait for `releasePromise` to complete instead
        // `released` is set atomically to ensure we don't end up with two concurrent releases
        if (released.compareAndSet(false, true))
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 6]
```diff
@@ -58,33 +88,40 @@ trait Resource[+A] {
...
             .transformWith {
-              case Success(value) => releaseResource(value).flatMap(_ => releaseSubResources())
-              case Failure(_) => releaseSubResources()
+              case Success(value) =>
+                releaseResource(value).flatMap(_ => releaseSubResources()) // Release all
+              case Failure(_) =>
+                releaseSubResources() // Only sub-release as the future will take care of itself
             }
-            .transform(
+            .transform( // Then use set the promise as release completion future
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
            .transform( // Finally, complete `releasePromise` to allow other releases to complete
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 7]
```diff
@@ -94,39 +131,67 @@ object Resource {
                 exception
               },
             )
-        else
+        else // The release completion future is already the promise
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
        else // A release is already in progress or completed; we wait for that instead
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 8]
```diff
@@ -94,39 +131,67 @@ object Resource {
...
+    * Builds a [[Resource]] from a [[Future]] plus release logic.
+    */
   def apply[T](future: Future[T])(releaseResource: T => Future[Unit])(
       implicit executionContext: ExecutionContext
   ): Resource[T] =
     nest(future)(releaseResource, () => Future.successful(()))
 
   /**
-    * Useful to wrap a simple [[Future]] that doesn't need releasing in a [[Resource]]
+    * Wraps in a [[Resource]] a simple [[Future]] that doesn't need any releasing.
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
    * Wraps a simple [[Future]] in a [[Resource]] that doesn't need to be released.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 9]
```diff
@@ -94,39 +131,67 @@ object Resource {
                 exception
               },
             )
-        else
+        else // The release completion future is already the promise
           releasePromise.future
     }
 
+  /**
+    * Builds a [[Resource]] from a [[Future]] plus release logic.
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
    * Builds a [[Resource]] from a [[Future]] and some release logic.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 10]
```diff
@@ -94,39 +131,67 @@ object Resource {
...
+    * Produces a [[Resource]] that has already failed with a given exception.
+    */
   def failed[T](exception: Throwable)(implicit executionContext: ExecutionContext): Resource[T] =
     Resource.fromFuture(Future.failed(exception))
 
+  /**
+    * Sequences a [[TraversableOnce]] of [[Resource]]s into a [[Resource]] of the [[TraversableOnce]] of their values.
+    *
+    * @param seq The [[TraversableOnce]] of [[Resource]]s.
+    * @param bf The projection from a [[TraversableOnce]] of resources into a one of their values.
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]:
```suggestion
    * @param bf The projection from a [[TraversableOnce]] of resources into one of their values.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala [hunk 11]
```diff
@@ -12,40 +12,65 @@ import scala.concurrent.{ExecutionContext, Future}
...
 
+  /** @see [[Resource.withFilter()]] */
   def withFilter(p: A => Boolean)(implicit executionContext: ExecutionContext): ResourceOwner[A] =
     new ResourceOwner[A] {
       override def acquire()(implicit executionContext: ExecutionContext): Resource[A] =
         self.acquire().withFilter(p)
     }
 
+  /**
+    * Uses the acquired [[Resource]]'s value asynchronously.
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala]:
```suggestion
    * Acquire the [[Resource]]'s value, use it asynchronously, and release it afterwards.
```

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala [hunk 12]
```diff
@@ -12,40 +12,65 @@ import scala.concurrent.{ExecutionContext, Future}
...
+    * @param executionContext The asynchronous task execution engine.
+    * @tparam T The asynchronous computation's value type.
+    * @return The asynchronous computation's [[Future]].
+    */
   def use[T](behavior: A => Future[T])(implicit executionContext: ExecutionContext): Future[T] = {
     val resource = acquire()
     resource.asFuture
       .flatMap(behavior)
-      .transformWith {
+      .transformWith { // Release the resource no matter if the computation succeeds or not
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala]:
```suggestion
      .transformWith { // Release the resource whether the computation succeeds or not
```

**Comment by fabiotudone-da** [comment]
> This is awesome, thank you!
> 
> I have left a lot of suggestions but they're minor. Please take a look and verify you're happy with them, and also that they fit in the line length, link appropriately, etc.

My pleasure and thank you for reviewing the draft so fast! All your suggestions make a lot of sense: I'm going to apply them in a single commit (incl. the CHANGELOG ceremony), move the PR to ready and if you are happy with the result we can merge it.

I think these docs will probably spare some time to people not too familiar with monadic resource management (like me) and this in turn will end up promoting this library for wider adoption (and reducing duplicated code).

**Pull request merged**